Move to float32 coordinates

andydotxyz opened the pull request:
Move coordinates to float32.
Move text size as well.

Add new Vector type for mouse events (also float32s).
More interoperability between Size/Position/Vector.

### Checklist:
<!-- Please tick these as appropriate using [x] -->

- [x] Tests included.
- [x] Lint and formatter run with no errors.
- [x] Tests all pass.
- [x] Any breaking changes have a deprecation path or have been discussed.

andydotxyz commented:
> What's the purpose of IsZero()? We could just as well implement this as `func IsZero(v Vec2) bool`.

It was in there already for the `Size` and `Position` types, I did not want to remove it.
Overall I thought we were trying to avoid package exported functions like that? I can happily remove it from the interface, but given that all the types have it, it seemed logical to keep it there.

charlesdaniels commented:
I have no problem it it, it just seemed like something that could be done as a function that takes the interface instance as an arg. It's pretty trivial either way.

toaster reviewed (changes requested):
A couple of questions and opinions. See comments and:

* The alignment of colour channel sliders and submenu arrows looked better before, IMO. Should we adjust that?
* Unfortunately I don’t have the time to read all discussions on Slack. A PR should explain its intent or the decision it is based on, but this doesn’t. Why should we make this move?
* The removal of the deprecated `Union` would have been better in an own PR, IMO.
* The same goes for the new `Vector` type.
* At least one commit (058e998f11de58053b5c7b873281c3e0a873300c) contains stuff that belongs to earlier commits :/. That might look like nit picking but I think focused commits help a lot with the review process. That’s why I put a significant amount of time into diff review before creating a commit and into commit clean-up before creating a PR.

andydotxyz commented on canvas/text_test.go:
```diff
@@ -51,7 +51,7 @@ func TestText_Layout(t *testing.T) {
 			text:  "abc",
 			align: fyne.TextAlignLeading,
-			size:  fyne.NewSize(500, 100),
+			size:  fyne.NewSize(500, 101),
```
The text inside has `21` height. If I don't change that then the content alignment has `.5` offsets which led to confusing off-by-1 changes, this seemed cleaner.

andydotxyz commented on test/markup_renderer.go:
```diff
@@ -110,19 +110,19 @@ func (r *markupRenderer) setFloatPosAttr(attrs map[string]*string, name string,
 
-func (r *markupRenderer) setIntAttrWithDefault(attrs map[string]*string, name string, i int, d int) {
-	if i == d {
+func (r *markupRenderer) setIntAttrWithDefault(attrs map[string]*string, name string, i float32, d float32) {
+	if int(i) == int(d) {
```
Thanks, I missed the name for some reason

andydotxyz commented on test/markup_renderer.go:
```diff
@@ -174,7 +174,7 @@ func (r *markupRenderer) setScaleModeAttr(attrs map[string]*string, name string,
 
 func (r *markupRenderer) setSizeAttr(attrs map[string]*string, name string, size fyne.Size) {
-	value := fmt.Sprintf("%dx%d", size.Width, size.Height)
+	value := fmt.Sprintf("%dx%d", int(size.Width), int(size.Height))
```
That is 1 option. But it led to like 10'000 changes in the text, and the 1.95 vs 2 didn't seem like a helpful change.
If this is a renderer then such things would not be visible. The Gl renderer will similarly round to nearest pixel...

toaster replied:
Yes, but this is not a renderer but a test helper that shows what a renderer would have as input.
If the renderer accesses some kind of Retina display where it reports a resolution of 1280x800 while having 2560x1600 physical pixels at hand, it might decide to place 19.5 at a different pixel than 19.0.
But even more important: exact numbers give the developer the ability to verify the placement logic with higher accuracy.
I see that the change is huge and therefore I suggest that you create an issue for this change to be done later when #1669 has landed.

andydotxyz replied:
Good plan

andydotxyz replied:
Added #1672

andydotxyz commented on widget/button_internal_test.go:
```diff
@@ -99,20 +99,14 @@ func TestButtonRenderer_Layout_Stretch(t *testing.T) {
-		iconYOffset = (textHeight - theme.IconInlineSize()) / 2
-	}
 	minIconHeight := fyne.Max(theme.IconInlineSize(), textHeight)
 	assert.Equal(t, 50+theme.Padding()*3, render.icon.Position().X, "icon x")
-	assert.Equal(t, 50+theme.Padding()*2+iconYOffset, render.icon.Position().Y, "icon y")
+	assert.Equal(t, 50+theme.Padding()*2, render.icon.Position().Y, "icon y")
 	assert.Equal(t, theme.IconInlineSize(), render.icon.Size().Width, "icon width")
 	assert.Equal(t, minIconHeight, render.icon.Size().Height, "icon height")
 	assert.Equal(t, 50+theme.Padding()*4+theme.IconInlineSize(), render.label.Position().X, "label x")
-	assert.Equal(t, 50+theme.Padding()*2+labelYOffset, render.label.Position().Y, "label y")
+	assert.Equal(t, 50+theme.Padding()*2, render.label.Position().Y, "label y")
```
I think this is neither, the tests were inconsistent with the code. In label the icon is not adjusted by changing it's position but by using the label resize to cause it's content to move (I think).
This change was required because the `0` int difference between them before was ignored, but now it is `0.5` which the actual layout code did not set, but the test expected.

toaster replied:
I just had a look at the button layout implementation. Your change is okay but I think the button layout needs an overhaul. It’s a mix of lay out by `Layout` and explicit moves and resizes. But, of course, this is off topic for this PR.

andydotxyz commented on internal/driver/glfw/driver_test.go:
```diff
@@ -42,7 +42,7 @@ func Test_gLDriver_AbsolutePositionForObject(t *testing.T) {
 	c.setMenuOverlay(movl)
 	w.SetContent(content)
-	w.Resize(fyne.NewSize(200, 200))
+	w.Resize(fyne.NewSize(200, 199))
```
As above. Content is an odd number minSize, so this avoids floating point strangeness in the value checking.

andydotxyz commented on geometry.go:
```diff
@@ -1,5 +1,12 @@
 package fyne
 
+// Vec2 marks geometry types that can operate as a coordinate vector
```
https://github.com/fyne-io/fyne/wiki/Proposal:-Use-Vectors-to-Represent-Position,-Size,-Etc.

I will see if the other name suggestions work better

andydotxyz replied:
I don't think Coordinate is better than Vector. It is generic on purpose - it can refer to a coordinate, a size, a delta etc. It's just X and Y quantity of some sort, and I can't find a better name for it

andydotxyz replied:
Of course we could make `Vector` a more specific `XYDelta` (as that is what what it's been used for so far), and then rename `Vec2` as `Vector` or `Vector2`.
The rest of the naming is now tidied up.

toaster replied:
Okay. I like the idea of using `Vector` for the interface.
The mathematical correct name for the current `Vector` implementation would be `Displacement` (https://en.wikipedia.org/wiki/Displacement_(geometry)). Unless you or someone else has a better idea, I would favour this one.

andydotxyz replied:
> The mathematical correct name for the current `Vector` implementation would be `Displacement`

Good point, though not a commonly used phrase and has a very specific meaning whereas this is trying to be quite generic. Delta feels more commonly used, or we could even go simpler with `Move` - this will be alongside `Size` and `Position` so looking for something commonplace word.

toaster replied:
I’m okay with `Move` or `Movement` but don’t really like `Delta` :).

andydotxyz commented:
Answers to the questions above that did not relate to lines of code (answered inline):

* The alignment of colour channel sliders and submenu arrows looked better before, IMO. Should we adjust that?

This change does not intentionally move anything, it is a result of the rounding/truncating being removed.
If we want to make specific adjustments after this that is OK, but I did not want to go ahead and do that as part of the coordinate type change.

* Unfortunately I don’t have the time to read all discussions on Slack. 

By using `int` in a `HiDPI` environment you cannot actually make use of each pixel position / size - for example on macOS Retina if we move an element it will jump 2 pixels. on iOS and Android this gets worse as SCALE is 3, 3.5 and 4 - so we see moves jump up to 4 pixels. This change allows us to do "sub-pixel" on the scalable layer leading to more fine-grained output.

* The removal of the deprecated Union would have been better in an own PR, IMO.

Maybe, but this is a re-working of geometry, of which the Union is part, so it seemed related enough and avoided a separate, tiny, PR.

* The same goes for the new Vector type.

As above, I guess you could argue that, but it's all related.

At least one commit (058e998) contains stuff that belongs to earlier commits :/. That might look like nit picking but I think focused commits help a lot with the review process. That’s why I put a significant amount of time into diff review before creating a commit and into commit clean-up before creating a PR.

Apologies for this. As you can imagine this change took 2 full days to perform and polish. I suspected that unpicking, cherry-picking and reworking the commits would have taken another half day which didn't seem like it offered value for time.

toaster commented:
> This change does not intentionally move anything, it is a result of the rounding/truncating being removed.
> If we want to make specific adjustments after this that is OK, but I did not want to go ahead and do that as part of the coordinate type change.

Okay, I just filed #1674 for this.

>     * The removal of the deprecated Union would have been better in an own PR, IMO.
> 
> Maybe, but this is a re-working of geometry, of which the Union is part, so it seemed related enough and avoided a separate, tiny, PR.
> 
>     * The same goes for the new Vector type.
> 
> As above, I guess you could argue that, but it's all related.

I understand your argument but nevertheless disagree. Both of these changes make complete sense on their own, without the `int`->`float32` transformation.
I understand that such things often happen as by-product of some development. I experience this on my own nearly every day. Nevertheless, it is a good idea to separate these into own PRs (even if they are tiny). This helps to keep the original PR focused and thus comprehensible and reviewable. You don’t do such commit optimisation for yourself (well, maybe a little bit) but for the reviewers and at least I would greatly appreciate this.
Sorry, for being such a nuisance about this but I think it pays out in the end by making the review process as smooth as possible.

> Apologies for this. As you can imagine this change took 2 full days to perform and polish. I suspected that unpicking, cherry-picking and reworking the commits would have taken another half day which didn't seem like it offered value for time.

Apology accepted, subterfuge not.
Big changes take time and often create a lot of stress. I know this and have to discipline myself over and over to not rush out things just to be done with them. Rest assured that such improvements to your commits and PRs are worth the effort by saving you and the reviewer time during the review process.
There are tools that help to make these operations less expensive. I’m using GitUp for this but I think there are even better alternatives. Given the three things I complained about, I would expect that re-arranging them into own PRs would not take longer than two hours. Yes, time is valuable but sometime investing a little bit in one place pays out in another one.

Sorry if some of this seems harsh, I don't want to offend you but I wanted to make my point clear.

toaster commented:
P.S.: Despite my long rant, I don’t expect that you extract any of these things into a separate PR _now_. I just want to suggest this for future PRs.

toaster commented on geometry.go:
```diff
@@ -1,12 +1,12 @@
 }
 
-// Vector is a generic X, Y coordinate type
+// Vector is a generic X, Y coordinate or size representation.
```
And movement :).

andydotxyz commented:
> P.S.: Despite my long rant, I don’t expect that you extract any of these things into a separate PR _now_. I just want to suggest this for future PRs.

Thanks for the clarification, I was worried about lots of required changes until I read this :)

stuartmscott reviewed (commented):
Looks good. I like that we got rid of a bunch of casts, tho it is a shame there isn't a way to define 32bit float point literals without the cast.

The color slider looks more inline with the label, so I think this change actually improves it.

andydotxyz commented:
> Sorry if some of this seems harsh, I don't want to offend you but I wanted to make my point clear.

Not at all - points taken. Perhaps I am running too fast at some of this stuff.
Trying to coordinate A book launch and 2.0 at the same time was probably a bad idea.

toaster commented:
Because the discussion grow to some length, here’s a checklist of the remaining tasks from my wishlist:

- [ ] don’t send floats to a method for `int`; either,
    - drop `MarkupRenderer`’s `setIntAttrWithDefault` and use `setFloatAttrWithDefault` instead, or
    - cast _before_ calling `setIntAttrWithDefault` and add the move to `setFloatAttrWithDefault` task to #1672, or
    - rename `setIntAttrWithDefault` to `setSizeAttrWithDefault` (and still add a note to #1672)
- [ ] improve naming of vector types and interface

andydotxyz commented:
I think the Int naminng was fixed yesteday @toaster - I moved it to Size based name, so it's less generic.

The other naming is now updated too.
In the chat channel `Delta` with `DX` and `DY` made a lot of sense. Let me know what you think :)

andydotxyz commented on canvas/line.go:
```diff
@@ -24,8 +24,8 @@ type Line struct {
 func (l *Line) Size() fyne.Size {
-	return fyne.NewSize(int(math.Abs(float64(l.Position2.X)-float64(l.Position1.X))),
-		int(math.Abs(float64(l.Position2.Y)-float64(l.Position1.Y))))
+	return fyne.NewSize(float32(math.Abs(float64(l.Position2.X)-float64(l.Position1.X))),
```
Yes I think we could add a few more helpers building on the new types
